src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,295 @@
+                                  typed_dictionary_segment->lower_bound(typed_search_value));
+
+        if (search_value_id == EMPTY_RESULT) {
+          continue;
+        }
+
+        if (search_value_id == SELECT_EVERYTHING) {
+          for (auto cell_id = 0ul, typed_segment_size = attribute_value_ids->size(); cell_id < typed_segment_size;
+               ++cell_id) {
+            position_list.push_back(RowID{static_cast<ChunkID>(chunk_id), static_cast<ChunkOffset>(cell_id)});
+          }
+          continue;
+        }
+
+        // Compare each value_id in the attribute vector to the calculated compare_value_id with comparator
+        for (auto cell_id = 0ul, typed_segment_size = attribute_value_ids->size(); cell_id < typed_segment_size;
+             ++cell_id) {
+          if (dictionary_comparator(attribute_value_ids->get(cell_id), search_value_id)) {
+            position_list.push_back(RowID{static_cast<ChunkID>(chunk_id), static_cast<ChunkOffset>(cell_id)});
+          }
+        }
+      } else if (const auto typed_reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+                 typed_reference_segment != nullptr) {
+        // For ReferenceSegments we materialize each referenced value and compare it to the search value
+
+        auto comparator = _get_comparator<Type>(_scan_type);
+
+        auto input_position_list = typed_reference_segment->pos_list();
+        auto input_reference_table = typed_reference_segment->referenced_table();
+
+        for (const auto& row_id : *input_position_list) {
+          const auto dereferenced_value = type_cast<Type>(
+              (*(input_reference_table->get_chunk(row_id.chunk_id).get_segment(_column_id)))[row_id.chunk_offset]);
+          if (comparator(dereferenced_value, typed_search_value)) {
+            position_list.push_back(row_id);
+          }
+        }
+      }
+    }
+  });
+  return position_list;
+}
+
+std::shared_ptr<const Table> TableScan::_create_reference_output_table(const std::shared_ptr<const Table>& input_table,
+                                                                       const std::vector<RowID>& position_list,
+                                                                       ChunkOffset target_chunk_size,
+                                                                       ColumnCount column_count) {
+  auto resulting_table = std::make_shared<Table>(target_chunk_size);
+
+  Chunk chunk;
+  auto ref_position_list = std::make_shared<PosList>(std::move(position_list));
+
+  // Return a Table consisting of one Chunk with one ReferenceSegment per Column
+  // Each ReferenceSegments refers to all rows that were evaluated to true by the scan
+  for (auto current_column_id = static_cast<ColumnCount>(0), columns = input_table->column_count();
+       current_column_id < columns; current_column_id++) {
+    resulting_table->add_column(input_table->column_name(static_cast<ColumnID>(current_column_id)),
```

Comment:
Minor: do `static_cast<ColumnID>(current_column_id)` just once outside the function calls?